ruby/red-arrow/lib/arrow/table.rb

```diff
@@ -261,12 +261,35 @@ def slice(*args)
         expected_n_args = nil
         case args.size
         when 1
-          if args[0].is_a?(Integer)
+          case args[0]
+          when Integer
             index = args[0]
             index += n_rows if index < 0
             return nil if index < 0
             return nil if index >= n_rows
             return Record.new(self, index)
+          when Hash
+            hash = args[0]
+            if hash.size != 1
+              message = "wrong number of hash size (given #{hash.size}, expected 1)"
+              raise ArgumentError, message
+            end
+
+            key = hash.keys[0]
+            value = hash.values[0]
+            case value
+            when Integer, String, TrueClass, FalseClass
+              slicers << (Slicer.new(self).send(key) == value)
+            when Range
+              if value.end == nil
+                slicers << (Slicer.new(self).send(key) >= value.first)
+              else
+                slicers << (Slicer.new(self).send(key) <= value.last)
+              end
+            else
+              message = "hash value must be Integer, String, TrueClass, FalseClass, Range"
+              raise ArgumentError, message  
+            end
```

Comment:
How about using `Hash#each`?

```suggestion
            condition_pairs = args[0]
            slicer = Slicer.new(self)
            conditions = []
            condition_pairs.each do |key, value|
              case value
              when Range
                # TODO: Optimize "begin <= key <= end" case by missing "between" kernel
                # https://issues.apache.org/jira/browse/ARROW-9843
                unless value.begin.nil?
                  conditions = (slicer[key] >= value.begin)
                end
                unless value.end.nil?
                  if value.exclude_end?
                    conditions << (slicer[key] < value.end)
                  else
                    conditions << (slicer[key] <= value.end)
                  end
                end
              else
                conditions = (slicer[key] == value)
              end
            end
            slicers << conditions.inject(:&)
```

Reply:
Thanks for good suggestion!!

Reply:
memo

Arrow::Slicer::Condition has [&] function.
This function allow to use multi condition.

Reply:
memo 

If Range Object is endless, 
Range.last occur error (cannot get the last element of endless range), 
but Range.end return nil.